src/dot_parser/mod.rs

```diff
@@ -0,0 +1,316 @@
+        if a_list.as_rule() != Rule::a_list {
+            continue;
+        }
+        let tokens: Vec<_> = a_list.into_inner().collect();
+        let mut i = 0usize;
+        while i < tokens.len() {
+            let key = tokens[i].as_str().trim().to_string();
+            if i + 1 < tokens.len() {
+                let val = tokens[i + 1].as_str().trim().to_string();
+                map.insert(key, unquote_str(&val));
+                i += 2;
+            } else {
+                map.insert(key, String::new());
+                i += 1;
+            }
+        }
+    }
+    map
+}
+
+/// Extract the first inner token of node_id
+fn node_id_to_string(pair: pest::iterators::Pair<Rule>) -> String {
+    if let Some(child) = pair.into_inner().next() {
+        return unquote_str(child.as_str().trim());
+    }
+    String::new()
+}
+
+#[pyfunction]
+pub fn from_dot(py: Python<'_>, dot_str: &str) -> PyResult<PyObject> {
+    let pairs = DotParser::parse(Rule::graph_file, dot_str).map_err(|e| {
+        PyErr::new::<pyo3::exceptions::PyValueError, _>(format!("DOT parse error: {}", e))
+    })?;
+
+    // Detect directedness from a clone of the iterator so we don't consume it.
+    let mut is_directed = false;
+    for pair in pairs.clone() {
+        if pair.as_rule() != Rule::graph_file {
+            continue;
+        }
+        let mut inner = pair.into_inner();
+        let first = inner.next().unwrap();
+        let graph_type_str = if first.as_rule() == Rule::strict {
+            inner.next().unwrap().as_str()
+        } else {
+            first.as_str()
+        };
+        is_directed = graph_type_str == "digraph";
+        break;
+    }
+
+    build_graph_enum(py, pairs, is_directed)
+}
+
+fn build_graph_enum(
+    py: Python<'_>,
+    pairs: pest::iterators::Pairs<Rule>,
+    is_directed: bool,
+) -> PyResult<PyObject> {
+    let mut node_map: HashMap<String, NodeIndex> = HashMap::new();
+    let graph_attrs = PyDict::new(py);
+
+    let mut default_node_attrs: HashMap<String, String> = HashMap::new();
+    let mut default_edge_attrs: HashMap<String, String> = HashMap::new();
+
+    let mut node_attrs_map: HashMap<String, PyObject> = HashMap::new();
+
+    let mut graph = if is_directed {
+        DotGraph::new_directed()
+    } else {
+        DotGraph::new_undirected()
+    };
+
+    for pair in pairs {
+        if pair.as_rule() != Rule::graph_file {
+            continue;
+        }
+        let mut inner = pair.into_inner();
+        let first = inner.next().unwrap();
+        if first.as_rule() == Rule::strict {
+            inner.next();
+        }
+
+        for rest in inner {
+            if rest.as_rule() != Rule::stmt_list {
+                continue;
+            }
+
+            for stmt in rest.into_inner() {
+                match stmt.as_rule() {
+                    Rule::node_stmt => {
+                        let mut it = stmt.into_inner();
+                        let nid = it.next().unwrap();
```

Comment:
Same comment here. If you keep needing a conversion from pest errors to `PyResult`, make a helper closure and do a `.map_err` or something along those lines